File: src/312_Burst_Balloons.py
# 312_Burst_Balloons
# https://leetcode.cn/problems/burst-balloons/

# You are given n balloons, indexed from 0 to n - 1. Each balloon is painted with a number 
# on it represented by an array nums. You are asked to burst all the balloons.

# If you burst the ith balloon, you will get nums[i - 1] * nums[i] * nums[i + 1] coins. 
# If i - 1 or i + 1 goes out of bounds of the array, then treat it as if there is a balloon with a 1 painted on it.

# Return the maximum coins you can collect by bursting the balloons wisely.

 
# Example 1:
# Input: nums = [3,1,5,8]
# Output: 167
# Explanation:
# nums = [3,1,5,8] --> [3,5,8] --> [3,8] --> [8] --> []
# coins =  3*1*5    +   3*5*8   +  1*3*8  + 1*8*1 = 167

# Example 2:
# Input: nums = [1,5]
# Output: 10

# from typing import List
# from pprint import pprint
# 错误解答
# class Solution:
#     def maxCoins(self, nums: List[int]) -> int:
#         nums = [1] + nums + [1]
#         l = len(nums)
#         dp = [[1] * l for _ in range(l)]
        
#         pprint(dp)

#         for i in range(1, l-1):
#             dp[i-1][i] = nums[i-1] * nums[i] * nums[i+1]
        
#         pprint(dp)
        
#         for i in range(l-2):
#             for j in range(l):
#                 for k in range(i+1, j):
#                     dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + nums[i] * nums[j] * nums[k])
        
#         pprint(dp)
                    
#         return dp[0][-1]
        
        
        
from typing import List

class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        nums = [1] + nums + [1]
        l = len(nums)
        dp = [[0] * l for _ in range(l)]

        for gap in range(2, l):
            for i in range(l-gap):
                j = i + gap
                for k in range(i+1, j):
                    dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + nums[i] * nums[k] * nums[j])

        return dp[0][-1]

        
s = Solution()
ans = s.maxCoins([3,1,5,8])
print(ans)
        

File: src/312_Burst_Balloons.rs
// 312_Burst_Balloons
// https://leetcode.cn/problems/burst-balloons/description/

// 有 n 个气球，编号为0 到 n - 1，每个气球上都标有一个数字，这些数字存在数组 nums 中。

// 现在要求你戳破所有的气球。戳破第 i 个气球，你可以获得 nums[i - 1] * nums[i] * nums[i + 1] 枚硬币。 
// 这里的 i - 1 和 i + 1 代表和 i 相邻的两个气球的序号。如果 i - 1或 i + 1 超出了数组的边界，那么就当它是一个数字为 1 的气球。

// 求所能获得硬币的最大数量。

use::std::cmp::max;
impl Solution {
    pub fn max_coins(mut nums: Vec<i32>) -> i32 {
        let mut new_nums = vec![1];
        new_nums.append(&mut nums);
        new_nums.push(1);
        let l = new_nums.len();

        let mut dp = vec![vec![0; l]; l];

        for gap in 2..l {
            for i in 0..l-gap {
                let j = i + gap;
                for k in i+1..j {
                    dp[i][j] = max(dp[i][j], dp[i][k] + dp[k][j] + new_nums[i] * new_nums[k] * new_nums[j]);
                }
            }
        }

        dp[0][l-1]
    }
}